File: 0786-search-in-a-sorted-array-of-unknown-size/0786-search-in-a-sorted-array-of-unknown-size.cpp
/**
 * // This is the ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * class ArrayReader {
 *   public:
 *     int get(int index);
 * };
 */

class Solution {
public:
    int search(const ArrayReader& reader, int target) {
        // if(reader.get(0)==target) 
        //     return 0;
        // if(reader.get(1)==target) 
        //     return 1;

        int start = 0;
        int end = start+1;

        while(start <= end)
        {
            if(target > reader.get(end))
            {
                start = end;
                end = end * 2;
            }
            else if(target <= reader.get(end))
            {
                int mid = start + (end-start)/2;
                if(reader.get(mid) == target){
                    return mid;
                }
                else if(target < reader.get(mid)){
                    end = mid-1;
                }
                else{
                    start = mid+1;
                }
            }
        }
        return -1;
    }
};

File: 0786-search-in-a-sorted-array-of-unknown-size/0786-search-in-a-sorted-array-of-unknown-size.java
/**
 * // This is ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface ArrayReader {
 *     public int get(int index) {}
 * }
 */

class Solution {
    public int search(ArrayReader reader, int target) {
        int lo = 0;
        int hi = lo+1;

        while(lo <= hi){
            // Step 1: Find the appropriate `hi` where target could be
            if(target > reader.get(hi)){
                lo = hi;
                hi = hi*2; // Keep doubling `hi` until we find the right range
            }
            // Step 2: Perform binary search within the range [lo, hi]
            else if(target <= reader.get(hi))
            {
                int mid = lo + (hi-lo)/2;

                if(target == reader.get(mid)){
                    return mid;
                } 
                else if(target < reader.get(mid))
                {
                    hi = mid-1;
                } 
                else 
                {
                    lo = mid+1;
                }
            }
        }
        return -1;
    }
}